File: projects/embervm/control/lib/embervm/s3_client/sigv4.ex
defmodule Embervm.S3Client.SigV4 do
  @moduledoc false

  @algorithm "AWS4-HMAC-SHA256"
  @region "us-east-1"
  @service "s3"
  @unsigned_payload "UNSIGNED-PAYLOAD"

  @spec sign(atom(), String.t(), [{String.t(), String.t()}], String.t(), map(), DateTime.t()) ::
          [{String.t(), String.t()}]
  def sign(method, url, headers, body_sha, %{access_key_id: id, secret_access_key: secret}, %DateTime{} = now)
      when is_binary(id) and id != "" and is_binary(secret) and secret != "" do
    amz_date = Calendar.strftime(now, "%Y%m%dT%H%M%SZ")
    date = Calendar.strftime(now, "%Y%m%d")

    headers =
      headers
      |> put_header("x-amz-content-sha256", body_sha)
      |> put_header("x-amz-date", amz_date)

    signed_headers =
      (["host", "x-amz-content-sha256", "x-amz-date"] ++
         if(header(headers, "content-type") == "", do: [], else: ["content-type"]))
      |> Enum.sort()

    canonical = canonical_request(method, url, headers, signed_headers, body_sha)
    scope = "#{date}/#{@region}/#{@service}/aws4_request"
    string_to_sign = "#{@algorithm}\n#{amz_date}\n#{scope}\n#{sha256_hex(canonical)}"
    signature = signature(secret, date, @region, @service, string_to_sign)

    authorization =
      "#{@algorithm} Credential=#{id}/#{scope}, SignedHeaders=#{Enum.join(signed_headers, ";")}, Signature=#{signature}"

    put_header(headers, "authorization", authorization)
  end

  def sign(_method, _url, headers, _body_sha, _creds, _now), do: headers

  @doc false
  def canonical_request(method, url, headers, signed_headers, body_sha) do
    uri = URI.parse(url)

    canonical_headers =
      Enum.map_join(signed_headers, "", fn name ->
        value = if name == "host", do: uri.authority || "", else: header(headers, name)
        "#{name}:#{normalize(value)}\n"
      end)

    [
      method |> to_string() |> String.upcase(),
      canonical_uri(uri.path),
      canonical_query(uri.query),
      canonical_headers,
      Enum.join(signed_headers, ";"),
      body_sha
    ]
    |> Enum.join("\n")
  end

  @doc false
  def signature(secret, date, region, service, string_to_sign) do
    k_date = mac("AWS4" <> secret, date)
    k_region = mac(k_date, region)
    k_service = mac(k_region, service)
    k_signing = mac(k_service, "aws4_request")
    mac(k_signing, string_to_sign) |> Base.encode16(case: :lower)
  end

  def unsigned_payload, do: @unsigned_payload

  defp canonical_uri(nil), do: "/"
  defp canonical_uri(""), do: "/"

  defp canonical_uri(path) do
    path
    |> String.split("/", trim: false)
    |> Enum.map_join("/", fn segment -> segment |> URI.decode() |> URI.encode(&URI.char_unreserved?/1) end)
  end

  defp canonical_query(nil), do: ""
  defp canonical_query(""), do: ""

  defp canonical_query(query) do
    query
    |> URI.query_decoder()
    |> Enum.map(fn {key, value} -> {URI.encode(key, &URI.char_unreserved?/1), URI.encode(value, &URI.char_unreserved?/1)} end)
    |> Enum.sort()
    |> Enum.map_join("&", fn {key, value} -> "#{key}=#{value}" end)
  end

  defp header(headers, name) do
    case Enum.find(headers, fn {key, _value} -> String.downcase(to_string(key)) == name end) do
      nil -> ""
      {_key, value} -> to_string(value)
    end
  end

  defp put_header(headers, name, value) do
    [{name, value} | Enum.reject(headers, fn {key, _} -> String.downcase(to_string(key)) == name end)]
  end

  defp normalize(value), do: value |> String.split() |> Enum.join(" ")
  defp sha256_hex(value), do: :crypto.hash(:sha256, value) |> Base.encode16(case: :lower)
  defp mac(key, value), do: :crypto.mac(:hmac, :sha256, key, value)
end

File: projects/embervm/noded/store/sigv4.go
package store

import (
	"crypto/hmac"
	"crypto/sha256"
	"encoding/hex"
	"net/http"
	"net/url"
	"sort"
	"strings"
)

const (
	sigV4Algorithm  = "AWS4-HMAC-SHA256"
	sigV4Region     = "us-east-1"
	sigV4Service    = "s3"
	unsignedPayload = "UNSIGNED-PAYLOAD"
)

type credentials struct {
	accessKeyID     string
	secretAccessKey string
}

// SignRequest adds an S3 SigV4 signature when credentials were configured.
// With no credentials it is a strict no-op, including leaving headers alone.
func (s *Store) SignRequest(req *http.Request) error {
	return s.sign(req)
}

func (s *Store) sign(req *http.Request) error {
	if s == nil || s.credentials.accessKeyID == "" || s.credentials.secretAccessKey == "" {
		return nil
	}
	now := s.now().UTC()
	amzDate := now.Format("20060102T150405Z")
	date := now.Format("20060102")
	req.Header.Set("x-amz-content-sha256", unsignedPayload)
	req.Header.Set("x-amz-date", amzDate)

	signedSet := map[string]struct{}{"host": {}}
	for name := range req.Header {
		lower := strings.ToLower(name)
		if lower != "authorization" {
			signedSet[lower] = struct{}{}
		}
	}
	signedHeaders := make([]string, 0, len(signedSet))
	for name := range signedSet {
		signedHeaders = append(signedHeaders, name)
	}
	sort.Strings(signedHeaders)
	canonical := canonicalRequest(req, signedHeaders, unsignedPayload)
	scope := date + "/" + sigV4Region + "/" + sigV4Service + "/aws4_request"
	stringToSign := sigV4Algorithm + "\n" + amzDate + "\n" + scope + "\n" + sha256Hex(canonical)
	signature := calculateSignature(s.credentials.secretAccessKey, date, sigV4Region, sigV4Service, stringToSign)
	req.Header.Set("Authorization", sigV4Algorithm+" Credential="+s.credentials.accessKeyID+"/"+scope+
		", SignedHeaders="+strings.Join(signedHeaders, ";")+", Signature="+signature)
	return nil
}

func canonicalRequest(req *http.Request, signedHeaders []string, payloadHash string) string {
	var headers strings.Builder
	for _, name := range signedHeaders {
		value := ""
		if name == "host" {
			value = req.Host
			if value == "" {
				value = req.URL.Host
			}
		} else {
			value = req.Header.Get(name)
		}
		headers.WriteString(name)
		headers.WriteByte(':')
		headers.WriteString(strings.Join(strings.Fields(value), " "))
		headers.WriteByte('\n')
	}
	return req.Method + "\n" + canonicalURI(req.URL) + "\n" + canonicalQuery(req.URL) + "\n" +
		headers.String() + "\n" + strings.Join(signedHeaders, ";") + "\n" + payloadHash
}

func canonicalURI(u *url.URL) string {
	path := u.EscapedPath()
	if path == "" {
		return "/"
	}
	parts := strings.Split(path, "/")
	for i, part := range parts {
		decoded, err := url.PathUnescape(part)
		if err == nil {
			parts[i] = awsEncode(decoded)
		}
	}
	return strings.Join(parts, "/")
}

func canonicalQuery(u *url.URL) string {
	values := u.Query()
	keys := make([]string, 0, len(values))
	for key := range values {
		keys = append(keys, key)
	}
	sort.Slice(keys, func(i, j int) bool { return awsEncode(keys[i]) < awsEncode(keys[j]) })
	parts := make([]string, 0, len(keys))
	for _, key := range keys {
		encodedKey := awsEncode(key)
		encodedValues := append([]string(nil), values[key]...)
		if len(encodedValues) == 0 {
			encodedValues = []string{""}
		}
		sort.Slice(encodedValues, func(i, j int) bool { return awsEncode(encodedValues[i]) < awsEncode(encodedValues[j]) })
		for _, value := range encodedValues {
			parts = append(parts, encodedKey+"="+awsEncode(value))
		}
	}
	return strings.Join(parts, "&")
}

func awsEncode(value string) string {
	return strings.ReplaceAll(url.QueryEscape(value), "+", "%20")
}

func sha256Hex(value string) string {
	sum := sha256.Sum256([]byte(value))
	return hex.EncodeToString(sum[:])
}

func calculateSignature(secret, date, region, service, stringToSign string) string {
	kDate := hmacSHA256([]byte("AWS4"+secret), date)
	kRegion := hmacSHA256(kDate, region)
	kService := hmacSHA256(kRegion, service)
	kSigning := hmacSHA256(kService, "aws4_request")
	return hex.EncodeToString(hmacSHA256(kSigning, stringToSign))
}

func hmacSHA256(key []byte, value string) []byte {
	h := hmac.New(sha256.New, key)
	_, _ = h.Write([]byte(value))
	return h.Sum(nil)
}
